### PR TITLE
Add CLI for configuring synchronous mode

### PR DESCRIPTION
**Pull request by shi-su**
**- What I did**
Add CLI to configure synchronous mode
`config synchronous_mode <enable|disable>`

**- How I did it**
Add a CLI script that writes the configuration of synchronous mode into CONFIG_DB and informs users to restart swss to apply the configuration.

**- How to verify it**
I verified it by testing locally with a virtual switch.

**- Previous command output (if the output of a command-line utility has changed)**
N/A

**- New command output (if the output of a command-line utility has changed)**
Wrote enable/disable synchronous mode into CONFIG_DB, swss restart required to apply the configuration

### CODE REVIEW

File: scripts/sonic-sync-mode [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+def main():
+    parser=argparse.ArgumentParser(description=
+          'Configure enable or disable synchronous mode between orchagent and syncd',
+          formatter_class=argparse.RawTextHelpFormatter,
+          epilog=
+'''
+sudo needed for configuring synchronous mode
+swss restart required to apply the configuration
+Options to restart swss and apply the configuration:
+    1. config save
```
Comment (qiluo-msft) [inline scripts/sonic-sync-mode:14]: >config save [](start = 7, length = 11)

`config save` also has `-y`. Do you want using?

File: scripts/sonic-sync-mode [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+       config reload -y
+    2. systemctl restart swss
```
Comment (qiluo-msft) [inline scripts/sonic-sync-mode:16]: >systemctl restart swss [](start = 7, length = 22)

Do you need `config save` before restart?

File: scripts/sonic-sync-mode [hunk 3]
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/python
+import argparse
+from swsscommon import swsscommon
+
+def main():
+    parser=argparse.ArgumentParser(description=
+          'Configure enable or disable synchronous mode between orchagent and syncd',
```
Comment (qiluo-msft) [inline scripts/sonic-sync-mode:7]: >Configure enable [](start = 11, length = 16)

Enable

File: scripts/sonic-sync-mode [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    args = parser.parse_args()
+
+    if args.mode == 'enable' or args.mode == 'disable':
+        config_db = swsscommon.DBConnector('CONFIG_DB', 0, False)
+        table = swsscommon.Table(config_db, 'DEVICE_METADATA')
+        fvs = swsscommon.FieldValuePairs([('synchronous_mode', args.mode)])
+        table.set('localhost', fvs)
+        print('Wrote %s synchronous mode into CONFIG_DB, swss restart required to apply the configuration'%(args.mode))
+    else:
+        print('Invalid argument: expect either enable or disable')
```
Comment (qiluo-msft) [inline scripts/sonic-sync-mode:29]: >print [](start = 8, length = 5)

print to stderr, and return with non zero error code.

File: scripts/sonic-sync-mode [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        parser.print_help()
+
+if __name__ == '__main__':
+    main()
```
Comment (qiluo-msft) [inline scripts/sonic-sync-mode:33]: >main [](start = 4, length = 4)

Do you have a plan to integrate into `config` CLI? Need a design doc.

File: scripts/sonic-sync-mode (hunk 1 above)
Comment (shi-su) [reply scripts/sonic-sync-mode]: Added `-y` in the comment

File: scripts/sonic-sync-mode (hunk 2 above)
Comment (shi-su) [reply scripts/sonic-sync-mode]: This is not needed, updated the help message to clarify the usage.

File: scripts/sonic-sync-mode (hunk 3 above)
Comment (shi-su) [reply scripts/sonic-sync-mode]: Fixed.

File: scripts/sonic-sync-mode (hunk 4 above)
Comment (shi-su) [reply scripts/sonic-sync-mode]: Updated as prescribed.

File: scripts/sonic-sync-mode (hunk 5 above)
Comment (shi-su) [reply scripts/sonic-sync-mode]: Integrated this into `config` CLI.

File: config/main.py [hunk 6]
```diff
@@ -1289,6 +1289,31 @@ def hostname(new_hostname):
...
+@config.command('synchronous_mode')
+@click.argument('sync_mode', metavar='<enable|disable>', required=True)
+def synchronous_mode(sync_mode):
+    """ Enable or disable synchronous mode between orchagent and syncd \n
+        swss restart required to apply the configuration \n
+        Options to restart swss and apply the configuration: \n
+            1. Keep and apply configuration after config reload \n
+                config save -y \n
+                config reload -y \n
+            2. Discard configuration after config reload \n
```
Comment (qiluo-msft) [inline config/main.py:1304]: >Discard configuration after config reload [](start = 15, length = 41)

This line has nothing to do with next line.

**Review by qiluo-msft — COMMENTED** [review]
One minor comment issue remains.

File: config/main.py (hunk 6 above)
Comment (shi-su) [reply config/main.py]: Removed the irrelevant line and only kept the commands to restart swss.

**Review by qiluo-msft — DISMISSED** [review]

**Comment by shi-su** [comment]
retest this please

**Comment by lguohan** [comment]
need unit test

**Review by lguohan — CHANGES_REQUESTED** [review]
please add unit test

File: config/main.py [hunk 7]
```diff
@@ -1289,6 +1289,29 @@ def hostname(new_hostname):
...
+            2. systemctl restart swss
+    """
+    
+    if sync_mode == 'enable' or sync_mode == 'disable':
+        config_db = ConfigDBConnector()
+        config_db.connect()
+        config_db.mod_entry('DEVICE_METADATA' , 'localhost', {"synchronous_mode" : sync_mode})
+        click.echo("Wrote %s synchronous mode into CONFIG_DB, swss restart required to apply the configuration" % sync_mode)
+    else:
+        click.echo("Error: Invalid argument %s, expect either enable or disable" % sync_mode)
```
Comment (lguohan) [inline config/main.py:1312]: click.fail()

File: config/main.py [hunk 8]
```diff
@@ -1289,6 +1289,29 @@ def hostname(new_hostname):
...
+            1. config save -y \n
+               config reload -y \n
+            2. systemctl restart swss
+    """
+    
+    if sync_mode == 'enable' or sync_mode == 'disable':
+        config_db = ConfigDBConnector()
+        config_db.connect()
+        config_db.mod_entry('DEVICE_METADATA' , 'localhost', {"synchronous_mode" : sync_mode})
+        click.echo("Wrote %s synchronous mode into CONFIG_DB, swss restart required to apply the configuration" % sync_mode)
```
Comment (lguohan) [inline config/main.py:1310]: please be more specific about the message, "you need config save -y and reload -y"

Comment (shi-su) [reply config/main.py]: Added the options to restart swss in the message.

File: config/main.py (hunk 7 above)
Comment (shi-su) [reply config/main.py]: Thanks for the suggestion. click does not seem to have a function "fail". Using "BadParameter" instead.

**Review by qiluo-msft — APPROVED** [review]
LGTM

**Comment by shi-su** [comment]
@lguohan Could you please let me know if this looks ok?

**Review by lguohan — APPROVED** [review]

**Pull request merged**